pkg/dns/dnsproxy.go

```diff
@@ -0,0 +1,175 @@
+		}
+	}()
+
+	tcpServer := &dnsgo.Server{Listener: tcpListener, Addr: udpConn.LocalAddr().String(), Net: "tcp", Handler: p.handler}
+	go func() {
+		p.log.Info("starting TCP server")
+		if err := tcpServer.ActivateAndServe(); err != nil {
+			p.log.Error(err, "failed to start TCP server")
+		}
+	}()
+
+	// Watch for termination signal
+	sigChan := make(chan os.Signal, 1)
+	signal.Notify(sigChan, os.Interrupt, syscall.SIGTERM)
+
+	<-sigChan
+
+	udpServer.Shutdown()
+	tcpServer.Shutdown()
+
+	return nil
+}
+
+func getUpdateCacheFunc(log logr.Logger, cache *DNSCache) func(lookupTime time.Time, response *dnsgo.Msg) {
+	return func(lookupTime time.Time, response *dnsgo.Msg) {
+		if response.Response && response.Rcode == dnsgo.RcodeSuccess {
+			scopedLog := log.WithValues(reqIdLogField, response.Id)
+			if err := cache.Update(lookupTime, response); err != nil {
+				scopedLog.Error(err, "failed to update DNS cache")
+			} else {
+				scopedLog.Info("cache updated")
+			}
+		}
+	}
+}
+
+// bindToPort attempts to bind to port for both UDP and TCP
+func (p DNSProxy) bindToPort() (*net.UDPConn, *net.TCPListener, error) {
+	var err error
+	var listener net.Listener
+	var conn net.PacketConn
+	bindAddr := net.JoinHostPort("", strconv.Itoa(int(p.port)))
+
+	defer func() {
+		if err != nil {
+			if listener != nil {
+				listener.Close()
+			}
+			if conn != nil {
+				conn.Close()
+			}
+		}
+	}()
+
+	listener, err = listenConfig().Listen(context.Background(), "tcp", bindAddr)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	conn, err = listenConfig().ListenPacket(context.Background(), "udp", listener.Addr().String())
+	if err != nil {
+		return nil, nil, err
+	}
+
+	p.log.Info("DNS proxy bound to address", "address", bindAddr)
+
+	return conn.(*net.UDPConn), listener.(*net.TCPListener), nil
+}
+
+func listenConfig() *net.ListenConfig {
+	return &net.ListenConfig{
+		Control: func(network, address string, c syscall.RawConn) error {
+			var opErr error
+			err := c.Control(func(fd uintptr) {
+				opErr = transparentSetsockopt(int(fd))
+				if opErr == nil {
+					opErr = unix.SetsockoptInt(int(fd), unix.SOL_SOCKET, unix.SO_MARK, 0x0B00)
+				}
+				if opErr == nil {
+					opErr = unix.SetsockoptInt(int(fd), unix.SOL_SOCKET, unix.SO_REUSEADDR, 1)
+				}
+				if opErr == nil {
+					opErr = unix.SetsockoptInt(int(fd), unix.SOL_SOCKET, unix.SO_REUSEPORT, 1)
+				}
+			})
+			if err != nil {
+				return err
+			}
+
+			return opErr
+		}}
+}
+
+func transparentSetsockopt(fd int) error {
+	var err4, err6 error
+
+	err6 = unix.SetsockoptInt(fd, unix.SOL_IPV6, unix.IPV6_TRANSPARENT, 1)
+	if err6 == nil {
+		err6 = unix.SetsockoptInt(fd, unix.SOL_IPV6, unix.IPV6_RECVORIGDSTADDR, 1)
+	}
+	if err6 != nil {
+		return err6
+	}
+
+	err4 = unix.SetsockoptInt(fd, unix.SOL_IP, unix.IP_TRANSPARENT, 1)
+	if err4 == nil {
+		err4 = unix.SetsockoptInt(fd, unix.SOL_IP, unix.IP_RECVORIGDSTADDR, 1)
+	}
+	if err4 != nil {
+		return err4
+	}
+
+	return nil
+}
```

Comment:
We need at least a comment where this comes from.

Also we need to check if nftables + `unix.IP_TRANSPARENT` + VRFs is working as expected.
It's intended to allow sockets to receive packets with any destination address/port.
If this combination is working we should not have to use a DNAT rule!

Reply:
see http://gsoc-blog.ecklm.com/iptables-redirect-vs.-dnat-vs.-tproxy/ and http://gsoc-blog.ecklm.com/gsoc-2018-project-submission/

Reply:
We skipped the idea of transparent sockets and use plain DNAT and a fixed DNS upstream server.